Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py

```diff
@@ -397,67 +332,116 @@ def _foregroundWidths(self):
+        logargs['LogName'] = 'peak_position'
+        logargs['LogText'] = format(self.beamPos, '.13f')
+        logargs['NumberType'] = 'Double'
+        AddSampleLog(**logargs)
+        # Add foreground start and end workspace indices to the sample logs of ws.
+        hws = self._foregroundWidths()
+        self.beamPosIndex = int(numpy.rint(self.beamPos))
+        if self.beamPosIndex > 255:
+            self.beamPosIndex = 255
+            self.self.log().warning('Is it a monitor spectrum?')
+        sign = self._workspaceIndexDirection(ws)
+        startIndex = self.beamPosIndex - sign * hws[0]
+        endIndex = self.beamPosIndex + sign * hws[1]
+        if startIndex > endIndex:
+            endIndex, startIndex = startIndex, endIndex
+        logargs['LogName'] = common.SampleLogs.FOREGROUND_START
+        logargs['LogText'] = str(startIndex)
+        logargs['NumberType'] = 'Int'
+        AddSampleLog(**logargs)
+        logargs['LogName'] = common.SampleLogs.FOREGROUND_CENTRE
+        logargs['LogText'] = str(self.beamPosIndex)
+        AddSampleLog(**logargs)
+        logargs['LogName'] = common.SampleLogs.FOREGROUND_END
+        logargs['LogText'] = str(endIndex)
+        AddSampleLog(**logargs)
+        return ws
+
+    def _moveDetector(self, ws):
+        """Perform detector position correction for reflected beams only."""
+        # Write twoTheta to sample logs
+        # Add theta to the sample logs of ws.
+        if not self.getProperty(Prop.BEAM_ANGLE).isDefault:
+            twoTheta = self.getProperty(Prop.BEAM_ANGLE).value
+        else:
+            spectrum_info = ws.spectrumInfo()
+            twoTheta = numpy.rad2deg(spectrum_info.twoTheta(self.beamPosIndex))
+        logargs = {'Workspace': ws,
+                   'LogType': 'Number',
+                   'EnableLogging': self._subalgLogging}
+        logargs['LogName'] = 'twoTheta'
+        logargs['LogText'] = format(twoTheta, '.13f')
+        logargs['NumberType'] = 'Double'
+        logargs['LogUnit'] = 'Degrees'
+        AddSampleLog(**logargs)
+        if self.getProperty(Prop.BEAM_ANGLE).isDefault:
+            return ws
+        # Move detector
+        detectorMovedWSName = self._names.withSuffix('detectors_moved')
+        self.instrumentName = ws.run().get('instrument.name').value
+        SpecularReflectionPositionCorrect(InputWorkspace=ws,
+                                          OutputWorkspace=detectorMovedWSName,
+                                          TwoTheta=twoTheta,
+                                          DetectorCorrectionType='RotateAroundSample',
+                                          DetectorComponentName='detector',
+                                          DetectorFacesSample=True,
+                                          PixelSize=common.pixelSize(self.instrumentName),
+                                          DirectLinePosition=self.beamPos,
```

Comment:
`DirectLinePosition` should come from the direct beam workspace. We should also utilize `DirectLineWorkspace`, why is it not here?